### ISSUE TITLE
LED mode stacking

### ISSUE THREAD

**Issue by algernon** [keyboardio/Kaleidoscope-LEDControl#10]
Based on keyboardio/Kaleidoscope-LEDEffect-SolidColor#2, #9, and LED mode interaction issues between `ActiveModColor` and others, I have a proposal: instead of having LED modes that have near-exclusive control over the LEDs, and a few special plugins like `ActiveModColor` that try to play well with them (but sometimes fail at that), how about making it possible to have more than one LED mode active at any given time? Similar in spirit to how layers work.

This way, `ActiveModColor` and `NumLock` would have it easier, as they'd just set the LEDs they want, and leave the rest to the modes that came before them. You could even have RainbowWave, Stalker, Wavepool, and ActiveModColor all active at the same time!

For this to work, we need to have an array that lists the modes we want (we already have this), a bitmap that describes which ones are active (we can use an `uint16_t` for this), and change a few methods to iterate over the active ones instead of calling the appropriate function for just one.

At a next step, we can optimize the LED modes further so that we iterate over the whole keymap only once instead of potentially as many times as we have active modes...

With this setup, we could rename `LEDMode` methods to:

- `setup(void)` (or `begin`, depending on how we implement it): Called once when `Kaleidoscope.use()`-ing the plugin.
- `activate(void)`: Called whenever the mode is activated.
- `update(void)` (or `update(row, col)`): The brains of it, called every cycle to update the LEDs.
- `deactivate(void)`: Called whenever the mode is deactivated.

**Comment by cdisselkoen** [comment]
This sounds great! Would add a lot of clarity to LED interactions. 

However, I'm wondering how this is going to assuage Jesse's concerns in Kaleidoscope-LEDEffect-SolidColor#2.  Specifically how are we going to avoid refreshing the entire keyboard on every update() with Solid Color? 

Independently, I've already started on a method rename to fix #9. I'll outline my ideas there when I get a chance (I'm busy the next few hours). The method rename is mostly orthogonal to the idea of having multiple active layers. 

_Sent from my OnePlus ONEPLUS A3000 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by algernon** [comment]
> However, I'm wondering how this is going to assuage Jesse's concerns in keyboardio/Kaleidoscope-LEDEffect-SolidColor#2. Specifically how are we going to avoid refreshing the entire keyboard on every update() with Solid Color?

It won't, not directly. What will happen is that `SolidColor` and similar plugins will get a little bit slower. But if they are the only thing changing the LEDs, we already have code in `KeyboardioScanner` (or `Kaleidoscope-Hardware-Model01`, I forgot), which only updates a bank of LEDs if they changed. So with `SolidColor` being the only active one, we won't actually reflesh the LEDs needlessly, we'll just have a few more function calls until data is discarded.

**Comment by algernon** [comment]
I started working on this, and uncovered a few serious issues.

1. If we call `LEDMode.update()` for every mode, each cycle, that will potentially iterate over the keymap many times.
2. If we call `LEDMode.refreshAt()` only, and get rid of `.update()`, plugins like `AlphaSquareEffect` will have a much harder time.
3. If we call `LEDMode.update()` first, then `LEDMode.refreshAt()` for each key, that will lead to unintuitive interactions between plugins.

Also, stacking has a good chance of slowing things down *considerably*.

What I will try next, is to have `LEDMode.update()` called only, and tweak the few plugins that iterate over the whole keymap so that they will only do that if necessary, to speed up our main loop.

**Comment by algernon** [comment]
Another issue I have, is with `LEDControl.refreshAt()`: if it is supposed to be used from a `LEDMode.update()`, then it should only call `LEDMode.refreshAt()` for modes below the one that called it. This makes things a tad more complicated.

**Comment by algernon** [comment]
Perhaps LED mode stacking is not the best of ideas... perhaps we should stick to having one exclusive mode, and things like `ActiveModColor` would not be LED modes, but they would just call `refreshAt()`. This gets us halfway there: we can have overrrides that play well together with the underlying mode, but our code is not needlessly complicated, nor slow.

**Comment by algernon** [comment]
Closing this, at least for now, because the cost of implementing this is too high.
